In [94]:
p, q = 227, 373

In [95]:
n = lambda p, q: p * q

In [96]:
n(p,q)

84671

In [97]:
def extended_gcd(a, b):  
    if b == 0:  
        return a, 1, 0  
    else:  
        gcd, x1, y1 = extended_gcd(b, a % b)  
        x = y1  
        y = x1 - (a // b) * y1  
        return gcd, x, y   

In [98]:
def reverse_value(a, m, func):
    NOD, u, _ = func(a, m)
    if(NOD == 1):
        return u%m
    else:
        raise Exception

In [99]:
def finding_e_d(e, p, q):
    d = reverse_value(e, (p-1)*(q-1), extended_gcd)
    if(extended_gcd(e, (p-1)*(q-1))[0] == 1):
        return (e, d)
    else:
        raise Exception

In [100]:
# reverse_value(1, (p-1)*(q-1), extended_gcd)

In [101]:
text = "Все счастливые семьи похожи друг на друга каждая несчастная семья несчастлива посвоему"

In [102]:
table = {'А':10, 'Б':11, 'В':12, 'Г':13, 'Д':14, 'Е':15, 'Ж':16, 'З':17, 'И':18, 'Й':19, 'К':20, 'Л':21, 'М':22, 'Н':23, 'О':24, 'П':25, 'Р':26, 'С':27, 'Т':28, 'У':29, 'Ф':30, 'Х':31, 'Ц':32, 'Ч':33, 'Ш':34, 'Щ':35, 'Ъ':36, 'Ы':37, 'Ь':38, 'Э':39, 'Ю':40, 'Я':41, ' ': 99}
rev_table = {v: k for k, v in table.items()}

In [103]:
big_text = lambda t: str.upper(t)

In [104]:
#translated_text = lambda t, d: "".join([str(d[symbol]) for symbol in t])
translated_text = lambda t, d: "".join(f"{d[symbol]:02d}" for symbol in t)

In [105]:
txt = translated_text(big_text(text), table)

In [106]:
#def get_blocks(text, n):
#    blocks = []
#    temp_num = ""
#    for i in range(0, len(text)):
#        temp_num += text[i]
#        if(int(temp_num) > n and int(temp_num)%10 == 0):
#            blocks.append(int(temp_num[:-2]))
#            temp_num = temp_num[-2:]
#        elif(int(temp_num) > n):
#            blocks.append(int(temp_num[:-1]))
#            temp_num = temp_num[-1:]
#    blocks.append(int(temp_num))
#    return blocks

In [107]:
def get_blocks(digits, n):
    """
    Разбивает строку цифр на блоки < n, без ведущих нулей.
    Пример: "121027419..." → [12102, 7419, ...]
    """
    blocks = []
    i = 0
    while i < len(digits):
        # Макс. возможная длина блока — сколько цифр в (n-1)
        max_len = len(str(n - 1))
        
        # Пробуем от max_len до 1
        block_len = min(max_len, len(digits) - i)
        found = False
        
        while block_len >= 1:
            candidate = digits[i:i + block_len]
            
            # Проверка 1: не начинается с '0' (кроме случая "0" — но у тебя коды ≥10, так что не будет)
            if len(candidate) > 1 and candidate[0] == '0':
                block_len -= 1
                continue
            
            num = int(candidate)
            
            # Проверка 2: num < n
            if num < n:
                blocks.append(num)
                i += block_len
                found = True
                break
            
            block_len -= 1
        
        if not found:
            raise ValueError(f"Не удалось разбить цифры '{digits[i:]}' на блоки < {n}")
    
    return blocks

In [108]:
# get_blocks(txt, n(p,q))

In [109]:
def pow_mod(base, exp, mod):
    result = 1
    base = base % mod
    while exp > 0:
        if exp & 1:
            result = (result * base) % mod
        base = (base * base) % mod
        exp >>= 1
    return result

In [118]:
pow_mod(27282, 5, 84671)

62493

In [119]:
pow(27282, 5, 84671)

62493

In [110]:
# pow_mod(4, 2, 100) 4^2 mod 100

In [111]:
def cipher_decipher(messages, e, n):
    res = []
    for message in messages:
        res.append(pow_mod(message, e, n))
    return res

In [112]:
import random

In [113]:
def generate_keys(p, q, e_candidates):
    open_closed_keys = []
    for i in e_candidates:
        try:
            open_closed_keys.append(finding_e_d(i, p, q))
        except Exception:
            pass
    return open_closed_keys

In [114]:
generate_keys(p,q, [5, 17, 65537]) # отдает числа e и d

[(5, 33629), (17, 59345), (65537, 75785)]

In [115]:
import time
keys = []
while True:
    answer = str(input("Что вы хотите сделать? Сгенерировать ключи - введите 1; Шифрование - введите 2; Расшифрование - введите 3; Выход - введите 4"))
    if(answer == "1"):
        keys = generate_keys(p,q, [5, 17, 65537])
        for index, (opened, closed) in enumerate(keys):
            input(f"{index}-ая пара ключей: e = {opened}, d = {closed}. Жмите Enter для продолжения")
        time.sleep(1)
    elif(answer == "2"):
        if(len(keys) == 0):
            keys = generate_keys(p,q, [5, 17, 65537])
            input(f"Поскольку вы не сгенерировали ключи - мы сгенерировали их за вас: {keys[0]}, {keys[1]}, {keys[2]}. Жмите Enter для продолжения")
        while True:
            fine = True
            text = big_text(str(input("Введите свое сообщение (для выхода напишите 1): ")))
            if(text == "1"):
                input("Выход. Жмите Enter для продолжения")
                break
            if(len(text) < 1):
                continue
            for symbol in text:
                if(symbol not in table.keys()):
                    fine = False
                    break
            if fine == False:
               input(f"Вы ввели некорректное предложение (в предложении присутствуют такие символы, которые НЕ шифруются). Жмите Enter для продолжения")
               continue
            while True:
                try:
                    chosen_keys = int(input(f"Выберите пару ключей (1-ое число - открытый ключ, а 2-ое - закрытый ключ): 1 - {keys[0]}, 2 - {keys[1]}, 3 - {keys[2]}"))
                    break
                except Exception:
                    input(f"Вы ввели НЕ число. Жмите Enter для продолжения")
                    continue
            #print(translated_text(text, table))
            print(get_blocks(translated_text(text, table), n(p,q)))
            cipher_nums = cipher_decipher(get_blocks(translated_text(text, table), n(p,q )), keys[chosen_keys-1][0], n(p,q))
            result = " ".join(str(num) for num in cipher_nums)
            print(result)
            break
    elif(answer == "3"):
        if(len(keys) == 0):
            keys = generate_keys(p,q, [5, 17, 65537])
            input(f"Поскольку вы не сгенерировали ключи - мы сгенерировали их за вас: {keys[0]}, {keys[1]}, {keys[2]}. Жмите Enter для продолжения")
        while True:
            fine = True
            text = str(input("Введите свое зашифрованное сообщение (для выхода напишите 1): "))
            if(text == "1"):
                input("Выход. Жмите Enter для продолжения")
                break
            if(len(text) < 2):
                continue
            for symbol in text:
                if(symbol not in ['0','1','2','3','4','5','6','7','8','9', ' ']):
                    fine = False
                    break
            if fine == False:
               input(f"Вы ввели некорректное зашифрованное сообщение (в сообщении присутствуют НЕ цифры). Жмите Enter для продолжения")
               continue
            while True:
                try:
                    chosen_keys = int(input(f"Выберите пару ключей (1-ое число - открытый ключ, а 2-ое - закрытый ключ): 1 - {keys[0]}, 2 - {keys[1]}, 3 - {keys[2]}"))
                    break
                except Exception:
                    input(f"Вы ввели НЕ число. Жмите Enter для продолжения")
                    continue
            try:
                cipher_nums = cipher_decipher([int(x) for x in text.split(' ')], keys[chosen_keys-1][1], n(p,q))
                result = "".join(str(num) for num in cipher_nums)
                #print([result[i:i + 2] for i in range(0, len(result), 2)])
                final_message = "".join(rev_table[int(symbol)] for symbol in [result[i:i + 2] for i in range(0, len(result), 2)])
            except Exception:
                input("Зашифрованное сообщение, которые вы ввели невозможно перевести в текст (проблема в выборе пары ключей или проблема в самом шифре). Жмите Enter чтобы вернуться к записи шифра")
                continue
            print(final_message)
            break

    elif(answer == "4"):
        input("Выход. Жмите Enter для продолжения")
        break
            
            

               
            

[27282, 61522, 18281, 52738, 9923, 15991, 0, 9929, 27251, 53129, 9910, 9920, 9932, 15232, 32427, 28412, 29920, 24282, 42637, 15992, 42399, 14101, 528]
62493 66912 18928 8303 47569 61337 0 34220 32342 77969 75179 71348 61478 63318 13608 37452 55799 51747 54155 13812 52895 39281 22604


44481 82588 54887 33451 47427 53150 - зашифрованное сообщение "мама мыла раму" при открытом ключе 5.

[27282, 61522, 18281, 52738, 9923, 15991, 0, 9929, 27251, 53129, 9910, 9920, 9932, 15232, 32427, 28412, 29920, 24282, 42637, 15992, 42399, 14101, 528]

(5, 33629) - 62493 66912 18928 8303 47569 13812 0 34220 32342 77969 75179 71348 61478 63318 13608 37452 55799 51747 54155 13812 52895 39281 22604 (Стремитесь не к успеху, а к ценностям которые он дает)